### PR TITLE
Fix for issue #270

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
This is reworking an older fix which proved to be incomplete; this code does some copy propagation only in the control which does checksum updates. It takes advantage of the fact that we only allow very restricted code in this control - the only method that can be called is the checksum.get() method.

The code can now handle multiple checksum units. It is not foolproof; some hand-written code patterns can trip it - the foolproof solution should probably use an extern in the BMv2 simulator for checksum calculations.

(I really don't know why the changes to the tableKeyNames.cpp file are in this PR; I should have done a rebase.)

### CODE REVIEW

File: backends/bmv2/lower.cpp [hunk 1]
```diff
@@ -127,34 +127,221 @@ const IR::Node* LowerExpressions::postorder(IR::Concat* expression) {
...
+    }
+};
+
+struct VariableDefinitions {
+    std::map<cstring, const VariableWriters*> writers;
+    VariableDefinitions* clone() const {
+        auto result = new VariableDefinitions();
+        result->writers = writers;
+        return result;
+    }
```
Comment (ChrisDodd) [inline backends/bmv2/lower.cpp]: Seems cleaner to use the copy ctor here, so the body would be just

    return new VariableDefinitions(*this);

File: backends/bmv2/lower.cpp [hunk 2]
```diff
@@ -127,34 +127,221 @@ const IR::Node* LowerExpressions::postorder(IR::Concat* expression) {
...
+    VariableDefinitions* join(const VariableDefinitions* other) const {
+        auto result = clone();
+        for (auto e : other->writers)
+            result->writers.emplace(e);
+        return result;
```
Comment (ChrisDodd) [inline backends/bmv2/lower.cpp]: If `e.first` (the string) already has an entry in `writers`, then this won't actually do anything -- shouldn't it be unioning the sets in this case?  Or can it never happen (should probably add a BUG_CHECK if so)

File: backends/bmv2/lower.cpp [hunk 3]
```diff
@@ -127,34 +127,221 @@ const IR::Node* LowerExpressions::postorder(IR::Concat* expression) {
...
+    bool hasUses(const IR::AssignmentStatement* statement) const {
+        return haveUses.find(statement) != haveUses.end();
+    }
+};
+
+// See the SimpleCopyProp pass below for the context in which this
+// analysis is run.  We take advantage that some more complex code
+// patterns have already been eliminated.
+class Accesses : public Inspector {
+    PathSubstitutions* substitutions;
```
Comment (ChrisDodd) [inline backends/bmv2/lower.cpp]: Using `ControlFlowVisitor` here would make the special handling of IfStatement (and other control flow constructs) below unnecessary.  Using `P4WriteContext` would obviate the need for the special handling of assignments, as well as make the code (potentially) work for out params of method calls.

Comment (mbudiu-vmw) [inline backends/bmv2/lower.cpp]: You are right about this; however, we only have 1 example using P4WriteContext in the compiler, and it is currently not working correctly - this is issue #242. So I have resorted to using a more explicit approach. 

I am also relying on the fact that on this architecture you cannot call methods with out arguments in the checksum verification stage.

File: backends/bmv2/lower.cpp (hunk 2 above)
Comment (mbudiu-vmw) [inline backends/bmv2/lower.cpp]: This is a bug, I will submit a patch.

File: backends/bmv2/lower.cpp [hunk 4]
```diff
@@ -153,15 +153,20 @@ struct VariableWriters {
...
     VariableDefinitions* join(const VariableDefinitions* other) const {
         auto result = clone();
-        for (auto e : other->writers)
-            result->writers.emplace(e);
+        for (auto e : other->writers) {
+            auto prev = writers.find(e.first);
+            if (prev == writers.end())
+                result->writers[e.first] = e.second;
+            else
+                result->writers[e.first] = prev->second->join(e.second);
```
Comment (ChrisDodd) [inline backends/bmv2/lower.cpp]: This could be just

    result->writers[e.first] = result->writers[e.first]->join(e.second);

**Pull request merged**